cli/fathom_web/test/test_list.py

```diff
@@ -0,0 +1,176 @@
+from click.testing import CliRunner
+
+from ..commands.list import main as list_main
+
+
+def test_end_to_end(tmp_path):
+    """Test expected outcome when using all of the optional parameters"""
+    # Make temporary in_directory and base_dir directories
+    base_dir, in_directory = make_directories(tmp_path)
+
+    # Make HTML files in in_directory in two separate subdirectories
+    # so we can exercise the recursive option.
+    a1, a2, b1, b2 = make_html_files(in_directory)
+
+    # Make the out_file we will save the output to
+    out_file = (base_dir / 'out_file.txt')
+
+    # Run fathom-list
+    result = CliRunner().invoke(
+        list_main,
+        [
+            in_directory.as_posix(),
+            f'-b{base_dir.as_posix()}',
+            '-r',
+            f'-o{out_file.as_posix()}',
+        ],
+    )
```

Comment:
This is cool! TIL how to simulate a command for a CLI test.

Reply:
It took me a crazy long time to get this right. The first time I went through this I had to trace through the execution of click to find that you need to pass strings for the arguments. This time, I found out an extremely hard way that you cannot have spaces in the strings. So `f'-b{base_dir.as_posix()}'` cannot be `f'-b {base_dir.as_posix()}'`. click will not strip the whitespace, so you get an OS error about an invalid argument to `open()`, but when the error is reported, I didn't notice the extra whitespace in the message.